Apply transverse aperture to thick elements.

cemitch99 opened the pull request:
Adds the comparison of particles against the transverse aperture within thick elements, at the end of each space charge slice.  This will close #763 .

Note:  Need to verify that using the same name `Aperture` for the thin aperture element and for the mixin class will not cause conflicts.  Otherwise, one of these should be renamed.

- [x] add mixin class
- [x] add application to a drift (as a template)
- [x] address particle id
- [x] add a benchmark example
- [x] add element documentation
- [x] add Python support
- [x] add application to other elements
- [x] update "thin" Aperture element to use consistent naming
- [x] Inputs: make `aperture` constructor non-breaking with warnings
- [x] Python: `Aperture` constructor is now breaking, please rename `xmax` argument to `aperture_x` and likewise for `y`
- [x] add example documentation

cemitch99 committed:
Initial files for applying aperture to thick elements.

cemitch99 committed:
Update aperture.H

Add namespace amrex::literals

cemitch99 committed:
Add Python bindings for drift.

cemitch99 committed:
Update elements.cpp

Add Python class for aperture mixin.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on src/particles/elements/Drift.H:
```diff
@@ -121,6 +128,11 @@ namespace impactx
             pt = ptout;
 
+            // apply transverse aperture
+            if (m_xmax > 0 && m_ymax > 0) {
```
I would move the `if` into `mixin/aperture.H`.

cemitch99 committed:
Add draft example.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Move if conditional into mixin.

cemitch99 committed:
Add drift aperture input documentation.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on examples/aperture/input_aperture_thick.in:
```diff
@@ -0,0 +1,50 @@
+drift.ymax = 1.5e-3
+
+# work-around for https://github.com/ECP-WarpX/impactx/issues/499
+amrex.the_arena_is_managed = 1
```
This is not needed anymore (fixed) and should be removed from all examples:
```suggestion
```

I added a PR for all other files in #793

ax3l commented on examples/aperture/run_aperture_thick.py:
```diff
@@ -0,0 +1,66 @@
+
+# work-around for https://github.com/ECP-WarpX/impactx/issues/499
+pp_amrex = amr.ParmParse("amrex")
+pp_amrex.add("the_arena_is_managed", 1)
```
Please see above:
```suggestion
from impactx import ImpactX, distribution, elements
```

cemitch99 commented:
Note:  Windows test fail now, but all tests were previously passing.

ax3l commented on src/particles/elements/mixin/aperture.H:
```diff
@@ -0,0 +1,107 @@
+ *
+ * This file is part of ImpactX.
+ *
+ * Authors: Axel Huebl
```
```suggestion
 * Authors: Axel Huebl, Chad Mitchell
```

ax3l commented on src/particles/elements/mixin/aperture.H:
```diff
@@ -0,0 +1,107 @@
+/* Copyright 2022-2023 The Regents of the University of California, through Lawrence
```
```suggestion
/* Copyright 2022-2025 The Regents of the University of California, through Lawrence
```

ax3l commented on src/particles/elements/mixin/aperture.H:
```diff
@@ -0,0 +1,107 @@
+#ifndef IMPACTX_ELEMENTS_MIXIN_APERTURE_H
+#define IMPACTX_ELEMENTS_MIXIN_APERTURE_H
+
+#include "particles/ImpactXParticleContainer.H"
+
+#include <ablastr/constant.H>
+
+#include <AMReX_Math.H>
+#include <AMReX_Extension.H>
+#include <AMReX_REAL.H>
```
```suggestion
#include <AMReX_Extension.H>
#include <AMReX_GpuQualifiers.H>
#include <AMReX_Math.H>
#include <AMReX_Particle.H>
#include <AMReX_REAL.H>
```

ax3l commented on src/particles/elements/Drift.H:
```diff
@@ -50,12 +54,15 @@ namespace impactx
             amrex::ParticleReal dx = 0,
             amrex::ParticleReal dy = 0,
             amrex::ParticleReal rotation_degree = 0,
+            amrex::ParticleReal xmax = 0,
+            amrex::ParticleReal ymax = 0,
```
I wonder if for all potential elements, `xmax` and `ymax` are general enough argument names.

They do make sense for the `Aperture` element, but for all other thick elements we might need something more specific, e.g., `x_open` or `x_aperture`?

Another small change I thought we could do is use `std::optional<amrex::ParticleReal> ... = std::nullopt`. That way, we do not redefine `0` with a jump, which technically is a fully blocking aperture. The equivalent in Python is then `x_aperture = None`, which is clearly no aperture.

What do you think?

cemitch99 replied:
I like `x_aperture`, and I'll make that change.  Will look into using `std::optional`.

ax3l replied:
To get the value out of [std::optional](https://en.cppreference.com/w/cpp/utility/optional), call [operator*](https://en.cppreference.com/w/cpp/utility/optional/operator*) or [.value()](https://en.cppreference.com/w/cpp/utility/optional/value).

We will need to write the real value into `m_x_aperture` because we cannot (should not) use `std::optional` on GPU. For no value, I would maybe write a negative value into the member variable and only apply the aperture for >=0. That way, 0 is well-defined (blocking).

cemitch99 replied:
To clarify the last comment:  We're avoiding `std::optional` altogether?  If so, I actually prefer the existing approach, to apply the aperture when x_aperture > 0 and y_aperture >0.  Why?  1) Allowing for blocking with x_aperture = 0 requires changing the logic/math in the mixin class, 2) A thick element with zero aperture has no use case (eg, its action is the same as a thin screen for blocking), 3) Allowing this case means setting the default values of x_aperture, y_aperture to an arbitrarily-chosen negative number.

ax3l replied:
Ok, then we can keep it simple and document <= zero as no-aperture :+1:

ax3l commented on src/particles/elements/mixin/aperture.H:
```diff
@@ -0,0 +1,107 @@
+               }
+
+            }
+
+
```
```suggestion
```

ax3l commented on src/python/elements.cpp:
```diff
@@ -215,6 +215,21 @@ void init_elements(py::module& m)
+        .def(py::init<>(),
+             "Mixin class for lattice elements with a transverse aperture."
+        )
+        .def_property_readonly("xmax",
+            &elements::Aperture::xmax,
+            "horizontal aperture in m"
+        )
+        .def_property_readonly("ymax",
```
Naming: please see above.

ax3l commented on src/python/elements.cpp:
```diff
@@ -669,13 +684,17 @@ void init_elements(py::module& m)
              py::arg("dx") = 0,
              py::arg("dy") = 0,
              py::arg("rotation") = 0,
+             py::arg("xmax") = 0,
+             py::arg("ymax") = 0,
```
If we use `std::optional` then this changes to
```suggestion
             py::arg("xmax") = py::none(),
             py::arg("ymax") = py::none(),
```

cemitch99 committed:
Apply suggestions from code review

Small changes (excluding naming and std::optional).

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

cemitch99 committed:
Update Drift.H

Remove "maybe unused".

cemitch99 committed:
Change name xmax -> x_aperture...

cemitch99 committed:
Fix one missing xmax -> x_aperture in elements.cpp.

cemitch99 commented on docs/source/usage/parameters.rst:
```diff
@@ -135,6 +135,8 @@ Lattice Elements
             * ``<element_name>.dx`` (``float``, in meters) horizontal translation error
             * ``<element_name>.dy`` (``float``, in meters) vertical translation error
             * ``<element_name>.rotation`` (``float``, in degrees) rotation error in the transverse plane
+            * ``<element_name>.x_aperture`` (``float``, in meters) horizontal half-aperture (elliptical)
+            * ``<element_name>.y_aperture`` (``float``, in meters) vertical half-aperture (elliptical)
```
```suggestion
            * ``<element_name>.aperture_x`` (``float``, in meters) horizontal half-aperture (elliptical)
            * ``<element_name>.aperture_y`` (``float``, in meters) vertical half-aperture (elliptical)
```

cemitch99 replied:
I like placing the coordinate (x,y) after "aperture", since we use a similar notation elsewhere.  For example `position_x`, `emittance_x`, `repeat_x` (in the Aperture element), etc.

cemitch99 commented on docs/source/usage/python.rst:
```diff
@@ -570,11 +570,16 @@ This module provides elements for the accelerator lattice.
    :param name: an optional name for the element
 
-.. py:class:: impactx.elements.Drift(ds, dx=0, dy=0, rotation=0, nslice=1, name=None)
+.. py:class:: impactx.elements.Drift(ds, dx=0, dy=0, rotation=0, x_aperture=0, y_aperture=0, nslice=1, name=None)
```
```suggestion
.. py:class:: impactx.elements.Drift(ds, dx=0, dy=0, rotation=0, aperture_x=0, aperture_y=0, nslice=1, name=None)
```

cemitch99 commented on docs/source/usage/python.rst:
```diff
@@ -570,11 +570,16 @@ This module provides elements for the accelerator lattice.
+   :param dx: horizontal translation error in m
+   :param dy: vertical translation error in m
+   :param rotation: rotation error in the transverse plane [degrees]
+   :param x_aperture: horizontal half-aperture (elliptical) in m
+   :param y_aperture: vertical half-aperture (elliptical) in m
```
```suggestion
   :param aperture_x: horizontal half-aperture (elliptical) in m
   :param aperture_y: vertical half-aperture (elliptical) in m
```

cemitch99 commented on examples/aperture/input_aperture_thick.in:
```diff
@@ -0,0 +1,47 @@
+
+drift.type = drift
+drift.ds = 0.123
+drift.x_aperture = 1.0e-3
+drift.y_aperture = 1.5e-3
```
```suggestion
drift.aperture_x = 1.0e-3
drift.aperture_y = 1.5e-3
```

cemitch99 commented on examples/aperture/run_aperture_thick.py:
```diff
@@ -0,0 +1,61 @@
+sim.lattice.extend(
+    [
+        monitor,
+        elements.Drift(name="drift", ds=0.123, x_aperture=1.0e-3, y_aperture=1.5e-3),
```
```suggestion
        elements.Drift(name="drift", ds=0.123, aperture_x=1.0e-3, aperture_y=1.5e-3),
```

cemitch99 commented on src/initialization/InitElement.cpp:
```diff
@@ -93,6 +93,28 @@ namespace detail
+    std::map<std::string, amrex::ParticleReal>
+    query_aperture (amrex::ParmParse& pp_element)
+    {
+        amrex::ParticleReal x_aperture = 0;
+        amrex::ParticleReal y_aperture = 0;
+        pp_element.query("x_aperture", x_aperture);
+        pp_element.query("y_aperture", y_aperture);
```
```suggestion
        amrex::ParticleReal aperture_x = 0;
        amrex::ParticleReal aperture_y = 0;
        pp_element.query("aperture_x", aperture_x);
        pp_element.query("aperture_y", aperture_y);
```

cemitch99 commented on src/initialization/InitElement.cpp:
```diff
@@ -93,6 +93,28 @@ namespace detail
+        pp_element.query("y_aperture", y_aperture);
+
+        std::map<std::string, amrex::ParticleReal> values = {
+                {"x_aperture", x_aperture},
+                {"y_aperture", y_aperture}
```
```suggestion
                {"aperture_x", aperture_x},
                {"aperture_y", aperture_y}
```

cemitch99 commented on src/initialization/InitElement.cpp:
```diff
@@ -128,8 +150,9 @@ namespace detail
+            auto b = detail::query_aperture(pp_element);
 
-            m_lattice.emplace_back( Drift(ds, a["dx"], a["dy"], a["rotation_degree"], nslice, element_name) );
+            m_lattice.emplace_back( Drift(ds, a["dx"], a["dy"], a["rotation_degree"], b["x_aperture"], b["y_aperture"], nslice, element_name) );
```
```suggestion
            m_lattice.emplace_back( Drift(ds, a["dx"], a["dy"], a["rotation_degree"], b["aperture_x"], b["aperture_y"], nslice, element_name) );
```

cemitch99 commented on src/particles/elements/Drift.H:
```diff
@@ -42,6 +44,8 @@ namespace impactx
          * @param dy vertical translation error in m
          * @param rotation_degree rotation error in the transverse plane [degrees]
+         * @param x_aperture horizontal half-aperture in m
+         * @param y_aperture vertical half-aperture in m
```
```suggestion
         * @param aperture_x horizontal half-aperture in m
         * @param aperture_y vertical half-aperture in m
```

cemitch99 commented on src/particles/elements/Drift.H:
```diff
@@ -50,12 +54,15 @@ namespace impactx
             amrex::ParticleReal dx = 0,
             amrex::ParticleReal dy = 0,
             amrex::ParticleReal rotation_degree = 0,
+            amrex::ParticleReal x_aperture = 0,
+            amrex::ParticleReal y_aperture = 0,
```
```suggestion
            amrex::ParticleReal aperture_x = 0,
            amrex::ParticleReal aperture_y = 0,
```

cemitch99 commented on src/particles/elements/Drift.H:
```diff
@@ -50,12 +54,15 @@ namespace impactx
           Thick(ds, nslice),
-          Alignment(dx, dy, rotation_degree)
+          Alignment(dx, dy, rotation_degree),
+          Aperture(x_aperture, y_aperture)
```
```suggestion
          Aperture(aperture_x, aperture_y)
```

cemitch99 commented on src/particles/elements/mixin/aperture.H:
```diff
@@ -0,0 +1,103 @@
+
+        /** A finite-length element
+         *
+         * @param x_aperture horizontal half-aperture size in m
+         * @param y_aperture vertical half-aperture size in m
```
```suggestion
         * @param aperture_x horizontal half-aperture size in m
         * @param aperture_y vertical half-aperture size in m
```

cemitch99 commented on src/particles/elements/mixin/aperture.H:
```diff
@@ -0,0 +1,103 @@
+         * @param y_aperture vertical half-aperture size in m
+         */
+        Aperture (
+            amrex::ParticleReal x_aperture,
+            amrex::ParticleReal y_aperture
+        )
+        : m_x_aperture(x_aperture), m_y_aperture(y_aperture)
```
```suggestion
            amrex::ParticleReal aperture_x,
            amrex::ParticleReal aperture_y
        )
        : m_aperture_x(aperture_x), m_aperture_y(aperture_y)
```

cemitch99 commented on src/particles/elements/mixin/aperture.H:
```diff
@@ -0,0 +1,103 @@
+            using namespace amrex::literals; // for _rt and _prt
+
+            // skip aperture application if x_aperture <= 0 or y_aperture <= 0
+            if (m_x_aperture > 0 && m_y_aperture > 0) {
+
+               // scale horizontal and vertical coordinates
+               amrex::ParticleReal const u = x / m_x_aperture;
+               amrex::ParticleReal const v = y / m_y_aperture;
```
```suggestion
            if (m_aperture_x > 0 && m_aperture_y > 0) {

               // scale horizontal and vertical coordinates
               amrex::ParticleReal const u = x / m_aperture_x;
               amrex::ParticleReal const v = y / m_aperture_y;
```

cemitch99 commented on src/particles/elements/mixin/aperture.H:
```diff
@@ -0,0 +1,103 @@
+         * @return horizontal aperture size in m
+         */
+        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+        amrex::ParticleReal x_aperture () const
+        {
+            return m_x_aperture;
```
```suggestion
        amrex::ParticleReal aperture_x () const
        {
            return m_aperture_x;
```

cemitch99 commented on src/particles/elements/mixin/aperture.H:
```diff
@@ -0,0 +1,103 @@
+         * @return vertical aperture size in m
+         */
+        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+        amrex::ParticleReal y_aperture () const
+        {
+            return m_y_aperture;
```
```suggestion
        amrex::ParticleReal aperture_y () const
        {
            return m_aperture_y;
```

cemitch99 commented on src/particles/elements/mixin/aperture.H:
```diff
@@ -0,0 +1,103 @@
+            return m_y_aperture;
+        }
+
+        amrex::ParticleReal m_x_aperture = 0; //! horizontal aperture size [m]
+        amrex::ParticleReal m_y_aperture = 0; //! vertical aperture size [m]
```
```suggestion
        amrex::ParticleReal m_aperture_x = 0; //! horizontal aperture size [m]
        amrex::ParticleReal m_aperture_y = 0; //! vertical aperture size [m]
```

cemitch99 commented on src/python/elements.cpp:
```diff
@@ -215,6 +215,21 @@ void init_elements(py::module& m)
+        .def(py::init<>(),
+             "Mixin class for lattice elements with a transverse aperture."
+        )
+        .def_property_readonly("x_aperture",
+            &elements::Aperture::x_aperture,
```
```suggestion
        .def_property_readonly("aperture_x",
            &elements::Aperture::aperture_x,
```

cemitch99 commented on src/python/elements.cpp:
```diff
@@ -215,6 +215,21 @@ void init_elements(py::module& m)
+            &elements::Aperture::x_aperture,
+            "horizontal aperture in m"
+        )
+        .def_property_readonly("y_aperture",
+            &elements::Aperture::y_aperture,
```
```suggestion
        .def_property_readonly("aperture_y",
            &elements::Aperture::aperture_y,
```

cemitch99 commented on src/python/elements.cpp:
```diff
@@ -669,13 +684,17 @@ void init_elements(py::module& m)
              py::arg("dx") = 0,
              py::arg("dy") = 0,
              py::arg("rotation") = 0,
+             py::arg("x_aperture") = 0,
+             py::arg("y_aperture") = 0,
```
```suggestion
             py::arg("aperture_x") = 0,
             py::arg("aperture_y") = 0,
```

cemitch99 committed:
Modify naming x_aperture -> aperture_x.

cemitch99 commented on src/initialization/InitElement.cpp:
```diff
@@ -128,8 +150,9 @@ namespace detail
+            auto b = detail::query_aperture(pp_element);
 
-            m_lattice.emplace_back( Drift(ds, a["dx"], a["dy"], a["rotation_degree"], nslice, element_name) );
+            m_lattice.emplace_back( Drift(ds, a["dx"], a["dy"], a["rotation_degree"], b["aperture_x"], b["y_aperture"], nslice, element_name) );
```
```suggestion
            m_lattice.emplace_back( Drift(ds, a["dx"], a["dy"], a["rotation_degree"], b["aperture_x"], b["aperture_y"], nslice, element_name) );
```

cemitch99 committed:
Update src/initialization/InitElement.cpp

cemitch99 committed:
Test workflow for adding elements using CFbend.

cemitch99 committed:
Merge branch 'development' into add_apertures_thick

cemitch99 commented:
Problems on Windows after merging in 'development'...
Appears to have resolved.

cemitch99 committed:
Add aperture input to all thick elements.

cemitch99 committed:
Update thin Aperture element naming.

cemitch99 committed:
Update element parameter docs.

cemitch99 committed:
Merge branch 'development' into add_apertures_thick

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on docs/source/usage/parameters.rst:
```diff
@@ -425,8 +455,8 @@ Lattice Elements
 
-            * ``<element_name>.xmax`` (``float``, in meters) maximum value of the horizontal coordinate
-            * ``<element_name>.ymax`` (``float``, in meters) maximum value of the vertical coordinate
+            * ``<element_name>.aperture_x`` (``float``, in meters) horizontal half-aperture (elliptical or rectangular)
+            * ``<element_name>.aperture_y`` (``float``, in meters) vertical half-aperture (elliptical or rectangular)
```
Oh, breaking change for existing `aperture` element users.

Not necessarily needed for this PR to work, but we can make it work by adding a bit more logic to transition.

cemitch99 replied:
Yes, I agree.  The case for consistent naming is pretty strong.  One of the main `aperture` users is  @SchroederSa, who was the first to suggest that `xmax` and `ymax` were not the best names.  Did you have in mind to support both input syntax by modifying `InitElement.cpp` and `python/elements.cpp`?

ax3l replied:
Yes, I'll add a fallback to make the transition easier there.

ax3l commented on docs/source/usage/python.rst:
```diff
@@ -963,12 +994,12 @@ This module provides elements for the accelerator lattice.
-.. py:class:: impactx.elements.Aperture(xmax, ymax, shape="rectangular", dx=0, dy=0, rotation=0, name=None)
+.. py:class:: impactx.elements.Aperture(aperture_x, aperture_y, shape="rectangular", dx=0, dy=0, rotation=0, name=None)
 
    A thin collimator element, applying a transverse aperture boundary.
 
-   :param xmax: maximum allowed value of the horizontal coordinate (meter)
-   :param ymax: maximum allowed value of the vertical coordinate (meter)
+   :param aperture_x: horizontal half-aperture (rectangular or elliptical) in m
+   :param aperture_y: vertical half-aperture (rectangular or elliptical) in m
```
See comment above: this would be a breaking change for the `Aperture` element. It is not needed to rename that one as well for this PR to work.

SchroederSa replied:
I like that description!

ax3l replied:
Thanks for checking ☺️

ax3l commented on src/initialization/InitElement.cpp:
```diff
@@ -386,13 +423,13 @@ namespace detail
-            pp_element.get("xmax", xmax);
-            pp_element.get("ymax", ymax);
+            pp_element.get("aperture_x", aperture_x);
+            pp_element.get("aperture_y", aperture_y);
```
If we want to rename the parameter (not needed for this PR), then I would add a `bool has_xmax = pp_element.query("xmax", xmax);` then if true, throw a warning and set the value for `aperture_x` (unless that one is set as well). Same for y.

But as I said, I don't think we need to rename the `Aperture` arguments with this PR.

ax3l replied:
Still have to add this part.

ax3l replied:
Added.

ax3l commented on src/particles/elements/Aperture.H:
```diff
@@ -58,8 +58,8 @@ namespace impactx
          * @param action specify action of domain (transmit/absorb)
-         * @param xmax maximum value of horizontal coordinate (m)
-         * @param ymax maximum value of vertical coordinate (m)
+         * @param aperture_x horizontal half-aperture (m)
+         * @param aperture_y vertical half-aperture (m)
```
See above: I think we could revert the changes to this file/element.

ax3l replied:
Discussed: we keep it, but make transition easier (warning and no immediate breakage).

ax3l commented on src/particles/elements/mixin/aperture.H:
```diff
@@ -0,0 +1,103 @@
+{
+    /** This is a helper class for applying a transverse aperture restriction to thick lattice elements
+     */
+    struct Aperture
```
There is a small collision risk of this mixin class colliding with the actual Aperture element. They are in the same namespace and have the exact same name.

We could consider renaming the mixin class to something else? (Very quick to do in my IDE if we have a good name.)

ax3l replied:
Renamed to `PipeAperture`. Wrapped in `impactx::elements::mixin` namespace as we do in Python.

ax3l commented on src/python/elements.cpp:
```diff
@@ -216,6 +216,20 @@ void init_elements(py::module& m)
         )
     ;
 
+    py::class_<elements::Aperture>(mx, "Aperture")
```
Surprised this does not cause trouble for us with our physical `"Aperture"` element. The names collide.

ax3l replied:
Avoided because of `elements.mixin.` namespace we recently added to Python, vs. `elements.` namespace for the physical ones.

Renamed now to `PipeAperature` to avoid confusion.

ax3l commented on src/python/elements.cpp:
```diff
@@ -216,6 +216,20 @@ void init_elements(py::module& m)
     ;
 
+    py::class_<elements::Aperture>(mx, "Aperture")
+        .def(py::init<>(),
+             "Mixin class for lattice elements with a transverse aperture."
+        )
```
I know we have it above for the other mixin classes, but looking at it we might want to check if we can remove this constructor for all of them, so one cannot create these mixin classes as a standalone variable, which would cause confusion.
```suggestion
```

ax3l replied:
Removed now for all mixin classes.

ax3l commented on src/python/elements.cpp:
```diff
@@ -331,10 +345,10 @@ void init_elements(py::module& m)
              }),
-             py::arg("xmax"),
-             py::arg("ymax"),
+             py::arg("aperture_x"),
+             py::arg("aperture_y"),
```
As above: we could avoid this breaking change for the `Aperture` element itself.

ax3l replied:
The break of API arg name here might be ok, for python people might not use the named arguments yet that much.

ax3l replied:
I'll keep this as breaking change for the Python API.

ax3l committed:
Rename Mixin Class: `PipeAperture`

ax3l committed:
Python: Mixin Elements w/o Constructor

We do not want to create variables from them.

ax3l committed:
Merge remote-tracking branch 'mainline/development' into add_apertures_thick

ax3l committed:
C++: Add `mixin::` Element Namespace

Mirror directory structure and separate from "physical"
elements as we do in Python.

ax3l committed:
Input Aperture: Backward Compatible Name

For inputs file.

ax3l committed:
Last Commit: Forgot Else

ax3l commented on src/particles/elements/Drift.H:
```diff
@@ -123,6 +130,9 @@ namespace impactx
             pt = ptout;
 
+            // apply transverse aperture
+            apply_aperture(x, y, idcpu);  //value of idcpu?
```
`idcpu` is passed above into the function, alongside the other particle attributes. All arguments passed to this function are set before the element is called.
```suggestion
            apply_aperture(x, y, idcpu);
```

ax3l committed:
Rename Mixin File: Pipe Aperture

cemitch99 committed:
Document benchmark example.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Remove comment re idcpu

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>
